### ISSUE TITLE
Open Sourcing Torsion

### ISSUE THREAD

**Issue by tomspilman** [GarageGames/Torque3D#1524]
Just want to give Torque developers a heads up.

We've been thinking about open sourcing [Torsion](https://www.garagegames.com/products/torsion) for years, but it was still pulling in a few hundred dollars a month... paying a few bills which helps a small indie like ourselves.

As of yesterday GarageGames is no longer selling Torsion, so rather than sell it ourselves the time has come to open source the thing.  I wanted to let people using T3D/T2D know about this and see if anyone was interested in helping manage the PRs that hopefully will bring it up to speed with the latest Torque releases.

I plan to get the repo up this weekend here:

 https://github.com/SickheadGames/Torsion

Also we did this with the [AdventureKit for T2D](https://github.com/SickheadGames/AdventureKit) a few years ago, but no one noticed.  Hoping this is more useful to people using Torque still.

**Comment by dottools** [comment]
Excellent. I think the [General Public License v3](http://www.gnu.org/licenses/gpl-3.0.en.html) might be suitable for Torsion as for licensing goes. What do you think?

**Comment by tomspilman** [comment]
Personally I've found GPL to be the worst open source license and something I actively avoid when looking for open source code.  I wouldn't want to poison the release with a license that keeps people from using the code.

I was thinking MIT, BSD, ZLIB, or MsPL...  but haven't researched the differences and pros/cons.  Torque3D/2D is MIT correct?

**Comment by Areloch** [comment]
I am beyond excited by this.

And yes, T2D and T3D are MIT.

**Comment by SteveYorkshire** [comment]
That's awfully nice of you all over at Sickhead.
grinning_eejit.jpg

**Comment by dottools** [comment]
Yes, Torque 3D and 2D are MIT licensed. It's pretty much the most straight forward (simple) open source license you could use. Only reason why I think GPLv3 would be suitable for Torsion is because it's a tool, not a library or middleware, like Blender. Only reason is because it forces whoever decides to fork the project and modify it and release binaries of it has to provide source code to the end-users, aka copy-left license. While MIT and others are are permissive / can pretty much do whatever you want and don't have to provide sources/changes.

**Comment by tomspilman** [comment]
Thanks @SteveYorkshire and @Areloch , but i still feel like a heel for not doing it sooner.

> I wouldn't be opposed to helping manage PRs if no one else does

I will try to setup at least an automated build to ensure the code still compiles after changes, but it will be up to others to make sure the change doesn't break features beyond that.  This is where it needs people familiar with T3D and T2D to help guide changes and development.

I would offer to help, but I'm so far removed from the Torque world now that I wouldn't know really.

> Only reason why I think GPLv3 would be suitable for Torsion is because it's a tool, not a library or middleware, like Blender. 

I see your point.  However I have had people interested in reusing the core of Torsion for other non-TorqueScript tools.  So I wouldn't want to handcuff those people beyond just making sure the original code base is credited somewhere.  If they don't want to release their code that is fine with me really.

**Comment by tomspilman** [comment]
Does anyone know the last version of Torsion we shipped and had for download on GG.com?  Was it v1.1.392?

**Comment by Lopuska** [comment]
Dear Tom,
This is an awesome announcement!
Thanks for this effort.
If you'll need help for PRs and other related work, I'll do my best, don't hesitate to ask me!

**Comment by tomspilman** [comment]
> If you'll need help for PRs and other related work, I'll do my best, don't hesitate to ask me!

I plan to put up a few general tasks up once I get the code in place like:

  - Fixes for latest T3D and T2D support.
  - Support latest VS C++ release.
  - Fix the old build script.
  - Update docs to reflect latest changes.
  - Upgrading wxWidgets.
  - Upgrading InnoSetup Installer.

... and more general stuff.  I sort of want to get the code base back up to working order before jumping into bigger tasks like UX improvements or Mac/Linux support.

**Comment by dottools** [comment]
Torsion downloaded from GG is reporting itself as v1.1.392 Final.

I wasn't aware that Torsion used wxWidgets, that's great because most of the hard work for cross platform support is already done in a nutshell.

**Comment by tomspilman** [comment]
> I wasn't aware that Torsion used wxWidgets, that's great because most of the hard work for cross platform support is already done in a nutshell.

Yep...  this is why I decided on wxWidgets way back in 2005.  We did a partial port to OSX in 2008, but never really finished the work.  I'm hoping that after upgrading to the latest wxWidgets version that project can be taken back up.

**Comment by eightyeight** [comment]
This is really awesome! Props to you guys for making this happen.

**Comment by tomspilman** [comment]
Code is up!

https://github.com/SickheadGames/Torsion

Writing up some issues that need to be addressed now!

**Comment by tomspilman** [comment]
Ok.... my issues are up:

 https://github.com/SickheadGames/Torsion/issues

Feel free to start dropping your own fix/feature requests up too.

Still the first goal is getting the code base back to a running state which requires we [update to the latest VS release](https://github.com/SickheadGames/Torsion/issues/7) and [replace our XML parser](https://github.com/SickheadGames/Torsion/issues/1).

**Comment by tomspilman** [comment]
@dottools and @practicing01...  I see you forked the code.  Please note in an issue what you plan to work on so that no one does any duplicate work.

**Comment by tomspilman** [comment]
Added a contributors guide to make sure people know what we expect:

 https://github.com/SickheadGames/Torsion/blob/master/CONTRIBUTING.md

And just setup a Gitter room for any live discussion needed:

 https://gitter.im/SickheadGames/Torsion

Looking forward to Torsion getting the love it needs. :)

**Comment by elfprince13** [comment]
This is awesome, thanks :)

btw, for future reference, while I think MIT is fine, LGPL is another good option: it disables the linking clauses of the GPL, so it won't spread to projects that embed it, but still requires people to release their changes to the package itself if they modify + release it in some form.

**Comment by mckl86** [comment]
ok so I am trying to download torsion (for 3d game programming book all in one) it only has 25 trial days then it asks for registration code or to buy it. Clicking on to buy it sends me to garagegames website where online store has been shutdown. What to do to get license? or registration code

**Comment by Areloch** [comment]
I haven't gotten around to finishing updating the torsion repo to pull out the really old bits and get 'er up to date. Hopefully I can get that sorted for 4.0.

In the meantime, I did compile the current open sourced build(as some bits of it require fussing) and put a zip up for people to nab [here](http://ghc-games.com/SC/Torsion.zip)
